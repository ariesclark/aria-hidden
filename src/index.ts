export type Undo = () => void;

const getDefaultParent = (originalTarget: Element | Element[]) => {
  if (typeof document === 'undefined') {
    return null;
  }

  const sampleTarget = Array.isArray(originalTarget) ? originalTarget[0] : originalTarget;
  return sampleTarget.ownerDocument.body;
};

let counterMap = new WeakMap<Element, number>();
let uncontrolledNodes = new WeakMap<Element, boolean>();
let markerMap: Record<string, WeakMap<Element, number>> = {};
let lockCount = 0;

/**
 * Marks everything except given node(pr nodes) as aria-hidden
 * @param {Element | Element[]} originalTarget - elements to keep on the page
 * @param parentNode - top element, defaults to document.body
 * @return {Undo} undo command
 */
export const hideOthers = (originalTarget: Element | Element[], parentNode = getDefaultParent(originalTarget), markerName = "data-aria-hidden"): Undo => {
  const targets = Array.isArray(originalTarget) ? originalTarget : [originalTarget];

  if (!markerMap[markerName]) {
    markerMap[markerName] = new WeakMap();
  }
  const markerCounter = markerMap[markerName];
  const hiddenNodes: Element[] = [];

  const elementsToKeep = new Set<Node>();
  const keep = ((el:Node | undefined) =>{
    if(!el || elementsToKeep.has(el)){
      return;
    }
    elementsToKeep.add(el);
    keep(el.parentNode);
  });
  targets.forEach(keep)

  const deep = (parent: Element | null) => {
    if (!parent || targets.indexOf(parent) >= 0) {
      return;
    }

    Array.prototype.forEach.call(parent.children, (node: Element) => {
      if (elementsToKeep.has(node)) {
        deep(node);
      } else {
        const attr = node.getAttribute('aria-hidden');
        const alreadyHidden = attr !== null && attr !== 'false';
        const counterValue = (counterMap.get(node) || 0) + 1;
        const markerValue = (markerCounter.get(node) || 0) + 1;

        counterMap.set(node, counterValue);
        markerCounter.set(node, markerValue);
        hiddenNodes.push(node);

        if (counterValue === 1 && alreadyHidden) {
          uncontrolledNodes.set(node, true);
        }

        if (markerValue === 1) {
          node.setAttribute(markerName, 'true');
        }

        if (!alreadyHidden) {
          node.setAttribute('aria-hidden', 'true')
        }
      }
    })
  };

  deep(parentNode);

  lockCount++;

  return () => {
    hiddenNodes.forEach(node => {
      const counterValue = counterMap.get(node) - 1;
      const markerValue = markerCounter.get(node) - 1;

      counterMap.set(node, counterValue);
      markerCounter.set(node, markerValue);

      if (!counterValue) {
        if (!uncontrolledNodes.has(node)) {
          node.removeAttribute('aria-hidden')
        }
        uncontrolledNodes.delete(node)
      }

      if (!markerValue) {
        node.removeAttribute(markerName);
      }
    });

    lockCount--;
    if (!lockCount) {
      // clear
      counterMap = new WeakMap();
      counterMap = new WeakMap();
      uncontrolledNodes = new WeakMap();
      markerMap = {};
    }
  }
};
