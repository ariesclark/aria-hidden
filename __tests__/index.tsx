import * as React from 'react';
import {mount} from 'enzyme';
import {hideOthers} from "../src";

describe('Specs', () => {
  const factory = () => {
    const parent = React.createRef<any>();
    const target1 = React.createRef<any>();
    const target2 = React.createRef<any>();
    const targetOutside1 = React.createRef<any>();
    const wrapper = mount(
      <div>
        <div ref={parent}>
          <div>hide me 1</div>
          <div ref={target1}>not me 2</div>
          <div>
            <div ref={target2}>not me 3</div>
          </div>
          <div ref={targetOutside1}>hide me 4</div>
          <svg>
            <text>svg</text>
          </svg>
          <div aria-hidden>I am already hidden! 5</div>
        </div>
        <div>dont touch me 6</div>
      </div>
    );
    const base = wrapper.html();

    return {
      base, wrapper,
      parent, target1, target2, targetOutside1,
    }
  };

  const getNearestAttribute = (node, name, parent) => {
    const attr = node.getAttribute(name);
    if (attr) {
      return attr;
    }
    if (node === parent || !node.parentNode) {
      return null;
    }
    return getNearestAttribute(node.parentNode, name, parent)
  }

  it('hides single', () => {
    const {
      base, parent, target1, target2, targetOutside1, wrapper
    } = factory();

    const unhide = hideOthers(target1.current, parent.current);
    expect(wrapper.update().html()).toMatchSnapshot();

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");

    unhide();
    expect(wrapper.html()).toEqual(base);
  });

  it('hides multiple', () => {
    const {
      base, parent, target1, target2, targetOutside1, wrapper
    } = factory();

    const unhide = hideOthers([target1.current, target2.current], parent.current);
    expect(wrapper.html()).toMatchSnapshot();

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");

    unhide();
    expect(wrapper.html()).toEqual(base);
  });

  it('hides cross', () => {
    const {
      base, parent, target1, target2, targetOutside1, wrapper
    } = factory();

    const unhide1 = hideOthers(target1.current, parent.current);
    expect(wrapper.html()).toMatchSnapshot();

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'data-aria-hidden', parent)).toBe("true");

    const unhide2 = hideOthers(target2.current, parent.current);

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'data-aria-hidden', parent)).toBe("true");

    expect(wrapper.html()).toMatchSnapshot();
    unhide1();

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'data-aria-hidden', parent)).toBe("true");

    expect(wrapper.html()).toMatchSnapshot();
    unhide2();

    expect(wrapper.html()).toEqual(base)
  });

  it('hides cross markers', () => {
    const {
      base, parent, target1, target2, targetOutside1, wrapper
    } = factory();

    const unhide1 = hideOthers(target1.current, parent.current, 'marker1');
    expect(getNearestAttribute(targetOutside1.current, 'marker1', parent)).toBe("true");

    const unhide2 = hideOthers(target2.current, parent.current, 'marker2');

    expect(getNearestAttribute(targetOutside1.current, 'marker1', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'marker2', parent)).toBe("true");

    unhide1();

    expect(getNearestAttribute(targetOutside1.current, 'marker1', parent)).toBe(null);

    expect(wrapper.html()).toMatchSnapshot();
    unhide2();

    expect(wrapper.html()).toEqual(base)
  });
  it('works on IE11', () => {
    // Simulate IE11 DOM Node implementation.
    HTMLElement.prototype.contains = Node.prototype.contains;
    delete Node.prototype.contains;
    const {
      base, parent, target1, target2, targetOutside1, wrapper
    } = factory();

    const unhide = hideOthers(target1.current, parent.current);
    expect(wrapper.update().html()).toMatchSnapshot();

    expect(getNearestAttribute(target1.current, 'aria-hidden', parent)).toBe(null);
    expect(getNearestAttribute(target2.current, 'aria-hidden', parent)).toBe("true");
    expect(getNearestAttribute(targetOutside1.current, 'aria-hidden', parent)).toBe("true");

    unhide();
    expect(wrapper.html()).toEqual(base);
  });
});
