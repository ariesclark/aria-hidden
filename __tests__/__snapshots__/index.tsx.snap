// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specs handles aria-live 1`] = `"<div><div><div><div>kept</div><div>hidden</div><div aria-live=\\"polite\\">not-hidden life</div><div aria-live=\\"off\\">hidden life</div></div><div>not-hidden</div></div></div>"`;

exports[`Specs hides cross 1`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div>not me 2</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs hides cross 2`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">not me 2</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs hides cross 3`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">not me 2</div><div><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs hides cross markers 1`] = `"<div><div><div aria-hidden=\\"true\\" marker2=\\"true\\">hide me 1</div><div marker2=\\"true\\" aria-hidden=\\"true\\">not me 2</div><div><div>not me 3</div></div><div aria-hidden=\\"true\\" marker2=\\"true\\">hide me 4</div><svg aria-hidden=\\"true\\" marker2=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" marker2=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs hides multiple 1`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div>not me 2</div><div><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs hides single 1`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div>not me 2</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;

exports[`Specs works on IE11 1`] = `"<div><div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 1</div><div>not me 2</div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><div>not me 3</div></div><div data-aria-hidden=\\"true\\" aria-hidden=\\"true\\">hide me 4</div><svg data-aria-hidden=\\"true\\" aria-hidden=\\"true\\"><text>svg</text></svg><div aria-hidden=\\"true\\" data-aria-hidden=\\"true\\">I am already hidden! 5</div></div><div>dont touch me 6</div></div>"`;
